### PR TITLE
Consolidate simulator

### PR DESCRIPTION
**Pull request by Modularius**
- Merges the functionality of run-simulator and simulator.
- Creates a unified language (using json) for simulating run control, metadata and data message, sending them to a Kafka broker in a prescribed sequence.
- Maintains the original functionality of `simulator ... single`, `simulator ... continuous` and `run-simulator ...`.
- Readme updated.

The component has been tested with the pipeline setup and running in the background, using the json file:
```json
{
    "voltage-transformation": {"scale": 1, "translate": 0 },
    "time-bins": 30000,
    "sample-rate": 1000000000,
    "digitiser-config": {
        "auto-digitisers": {
            "num-digitisers": 32,
            "num-channels-per-digitiser": 8
        }
    },
    "pulses": [
                {
                    "pulse-type": "flat",
                    "start":  { "random-type": "exponential", "lifetime": { "float": 2200 } },
                    "width":  { "random-type": "uniform", "min": { "float": 20 }, "max": { "float": 50 } },
                    "height": { "random-type": "uniform", "min": { "float": 30 }, "max": { "float": 70 } }
                },
                {
                    "pulse-type": "triangular",
                    "start":     { "random-type": "exponential", "lifetime": { "float": 2200 } },
                    "width":     { "random-type": "uniform", "min": { "float": 20 }, "max": { "float": 50 } },
                    "peak_time": { "random-type": "uniform", "min": { "float": 0.25 }, "max": { "float": 0.75 } },
                    "height":    { "random-type": "uniform", "min": { "float": 30 }, "max": { "float": 70 } }
                }],
    "event-lists": [
        {
            "pulses": [ {"weight": 1, "pulse-index": 0}, {"weight": 1, "pulse-index": 1} ],
            "noises": [],
            "num-pulses": { "random-type": "constant", "value": {"int": 50 }}
        }
    ],
    "schedule" : [
        { "set-timestamp": "now"},
        { "send-run-start" : { "name": "MyRun", "instrument": "MuSR" } },
        { "wait-ms": 100 },
        { "send-run-log-data" : {
            "source-name" : "A-Double-Value",
            "value-type" : "float64",
            "value" : ["0"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-Float-Value",
            "value-type" : "float32",
            "value" : ["-58"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-Byte-Value",
            "value-type" : "int8",
            "value" : ["-54"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-UByte-Value",
            "value-type" : "uint8",
            "value" : ["52"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-Short-Value",
            "value-type" : "int16",
            "value" : ["-540"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-UByte-Value",
            "value-type" : "uint16",
            "value" : ["524"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-Int-Value",
            "value-type" : "int32",
            "value" : ["-2234540"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-UInt-Value",
            "value-type" : "uint32",
            "value" : ["53524"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-Long-Value",
            "value-type" : "int64",
            "value" : ["-22111111111345400"]
        } },
        { "wait-ms": 100 },
        { "send-run-log-data" : {
            "source-name" : "A-Long-Value",
            "value-type" : "int64",
            "value" : ["-22111111111345400"]
        } },
        { "send-run-log-data" : {
            "source-name" : "A-ULong-Value",
            "value-type" : "uint64",
            "value" : ["53524000000000"]
        } },
        { "wait-ms": 100 },
        { "send-sample-env-log" : {
            "name" : "Earth's Core",
            "values-type" : "uint64",
            "location": "unknown",
            "values" : ["36", "36.1", "36.8"]
        } },
        { "send-sample-env-log" : {
            "name" : "Earth's Other Core",
            "values-type" : "uint64",
            "location": "unknown",
            "values" : ["36.9", "36.1", "36.8"]
        } },
        { "send-alarm" : {
            "source-name" : "Earth's Core",
            "severity" : "MINOR",
            "message": "fml"
        } },
        { "wait-ms": 100 },
        { "frame-loop" : {
                "start": 0,
                "end": 9,
                "schedule" : [
                    { "set-timestamp": { "advance-by-ms": 20}},
                    {
                        "digitiser-loop" : {
                            "start": 0,
                            "end": 31,
                            "schedule" : [
                                { "generate-trace": { "event-list-index": 0, "repeat": 8}},
                                { "send-digitiser-trace": "pop-front"}
                            ]
                        }
                    },
                    { "set-timestamp": { "advance-by-ms": 20}}
                ]
            }
        },
        { "wait-ms": 100 },
        { "set-timestamp": { "advance-by-ms": 200}},
        { "send-run-stop": { "name": "MyRun"} }
    ]
}
```

Closes https://github.com/STFC-ICD-Research-and-Design/supermusr-data-pipeline/issues/105.

### CODE REVIEW

**Review by DanNixon — COMMENTED** [review]
Couple of initial notes.

File: docs/simulator.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    Action ..|> EmitDigitiserEventList
+    class EmitFrameEventList {
+        Source trace_source
+    }
+    Action ..|> EmitFrameEventList
+    Action ..|> SetProperty
+```
+
+```mermaid
```
[the remark this answers is not in the document]
Comment (Modularius) [reply docs/simulator.md]: I should have got rid of this actually.

**Comment by DanNixon** [comment]
@Modularius There are a couple of conflicts to resolve.

**Review by DanNixon — CHANGES_REQUESTED** [review]
:+1: Looks good.

Could you just make sure that all members of a `struct` that implements `Parser` that are not a subcommand are documented as well as the struct itself.

This is important as this is where the clap generated CLI help text comes from.

File: simulator/src/runs/alarm.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+use anyhow::{anyhow, Result};
+use supermusr_streaming_types::ecs_al00_alarm_generated::Severity;
+
+pub(crate) fn severity(severity: &str) -> Result<Severity> {
+    match severity {
```
Comment (DanNixon) [inline simulator/src/runs/alarm.rs:5]: `clap` can parse enums directly, you should not have to do this.

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: simulator/src/runs/alarm.rs (hunk 2 above)
Comment (Modularius) [reply simulator/src/runs/alarm.rs]: Severity is not an enum, but I can use custom enums as discussed earlier.

**Review by DanNixon — APPROVED** [review]

**Pull request merged**